packages/aws-cdk/lib/api/cloudformation-deployments.ts

```diff
@@ -178,18 +186,85 @@ export class CloudFormationDeployments {
    * - SDK loaded with the right credentials for calling `CreateChangeSet`.
    * - The Execution Role that should be passed to CloudFormation.
    */
-  private async prepareSdkFor(stack: CloudFormationStackArtifact, roleArn?: string, mode = Mode.ForWriting) {
+  private async prepareSdkFor(stack: cxapi.CloudFormationStackArtifact, roleArn?: string, mode = Mode.ForWriting) {
     if (!stack.environment) {
       throw new Error(`The stack ${stack.displayName} does not have an environment`);
     }
+
     const resolvedEnvironment = await this.sdkProvider.resolveEnvironment(stack.environment);
 
-    const stackSdk = await this.sdkProvider.forEnvironment(stack.environment, mode);
+    // Substitute any placeholders with information about the current environment
+    const arns = await this.replaceEnvPlaceholders({
+      assumeRoleArn: stack.assumeRoleArn,
+
+      // Use the override if given, otherwise use the field from the stack
+      cloudFormationRoleArn: roleArn ?? stack.cloudFormationExecutionRoleArn,
+    }, resolvedEnvironment);
+
+    const stackSdk = arns.assumeRoleArn
+      ? await this.sdkProvider.withAssumedRole(arns.assumeRoleArn, undefined, resolvedEnvironment.region)
+      : await this.sdkProvider.forEnvironment(resolvedEnvironment, mode);
 
     return {
       stackSdk,
       resolvedEnvironment,
-      cloudFormationRoleArn: roleArn,
+      cloudFormationRoleArn: arns.cloudFormationRoleArn,
     };
   }
+
+  /**
+   * Replace the {ACCOUNT} and {REGION} placeholders in all strings found in a complex object.
+   */
+  private async replaceEnvPlaceholders<A extends { }>(object: A, env: cxapi.Environment): Promise<A> {
+    return cxapi.EnvironmentPlaceholders.replaceAsync(object, {
+      accountId: () => Promise.resolve(env.account),
+      region: () => Promise.resolve(env.region),
+      partition: async () => {
+        // We need to do a rather complicated dance here to get the right
+        // partition value to substitute into placeholders :(
+        const defaultAccount = await this.sdkProvider.defaultAccount();
+        return env.account === defaultAccount?.accountId
+          ? defaultAccount.partition
+          : (await (await this.sdkProvider.forEnvironment(env, Mode.ForReading)).currentAccount()).partition;
+      },
+    });
+  }
+
+  /**
+   * Publish all asset manifests that are referenced by the given stack
+   */
+  private async publishStackAssets(stack: cxapi.CloudFormationStackArtifact, bootstrapStack: ToolkitInfo | undefined) {
+    const stackEnv = await this.sdkProvider.resolveEnvironment(stack.environment);
+    const assetArtifacts = stack.dependencies.filter(isAssetManifestArtifact);
+
+    for (const assetArtifact of assetArtifacts) {
+      this.validateBootstrapStackVersion(stack.stackName, assetArtifact.requiresBootstrapStackVersion, bootstrapStack);
```

Comment:
duplicate?

Reply:
One is for the stack, the other for each individual asset.